// @clr/angular/esm5/popover/signpost/signpost.js
import * as tslib_1 from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component, ContentChild, ElementRef } from '@angular/core';
import { IfOpenService } from '../../utils/conditional/if-open.service';
import { POPOVER_HOST_ANCHOR } from '../common/popover-host-anchor.token';
import { ClrSignpostTrigger } from './signpost-trigger';
import { ClrCommonStrings } from '../../utils/i18n/common-strings.interface';
import { SignpostIdService } from './providers/signpost-id.service';
var ClrSignpost = /** @class */ (function () {
    function ClrSignpost(commonStrings) {
        this.commonStrings = commonStrings;
        /**********
         * @property useCustomTrigger
         *
         * @description
         * Flag used to determine if we need to use the default trigger or a user supplied trigger element.
         *
         */
        this.useCustomTrigger = false;
    }
    Object.defineProperty(ClrSignpost.prototype, "customTrigger", {
        /**********
         * @property signPostTrigger
         *
         * @description
         * Uses ContentChild to check for a user supplied element with the ClrSignpostTrigger on it.
         *
         */
        set: function (trigger) {
            this.useCustomTrigger = !!trigger;
        },
        enumerable: true,
        configurable: true
    });
    tslib_1.__decorate([
        ContentChild(ClrSignpostTrigger, { static: false }),
        tslib_1.__metadata("design:type", ClrSignpostTrigger),
        tslib_1.__metadata("design:paramtypes", [ClrSignpostTrigger])
    ], ClrSignpost.prototype, "customTrigger", null);
    ClrSignpost = tslib_1.__decorate([
        Component({
            selector: 'clr-signpost',
            template: "\n        <ng-container *ngIf=\"!useCustomTrigger\">\n            <button\n                type=\"button\"\n                class=\"signpost-action btn btn-small btn-link\"\n                clrSignpostTrigger>\n                <clr-icon shape=\"info\" [attr.title]=\"commonStrings.info\"></clr-icon>\n            </button>\n        </ng-container>\n        \n        <ng-content></ng-content>\n    ",
            host: { '[class.signpost]': 'true' },
            providers: [IfOpenService, { provide: POPOVER_HOST_ANCHOR, useExisting: ElementRef }, SignpostIdService]
        })
        /*********
         *
         * @class ClrSignpost
         *
         * @description
         * Class used to configure and control the state of a ClrSignpost and its associated ClrSignpostContent.
         * It supports the clrPosition with a 'right-middle' default.
         *
         */
        ,
        tslib_1.__metadata("design:paramtypes", [ClrCommonStrings])
    ], ClrSignpost);
    return ClrSignpost;
}());
export { ClrSignpost };
